fuse_variables/test/test_orientation_3d_stamped.cpp

```diff
@@ -0,0 +1,203 @@
+  {
+    Orientation3DStamped variable1(ros::Time(12345678, 910111213), uuid_1);
+    Orientation3DStamped variable2(ros::Time(12345678, 910111213), uuid_1);
+    EXPECT_EQ(variable1.uuid(), variable2.uuid());
+  }
+
+  // Verify two orientations with the same timestamp but different hardware IDs generate different UUIDs
+  {
+    Orientation3DStamped variable1(ros::Time(12345678, 910111213), uuid_1);
+    Orientation3DStamped variable2(ros::Time(12345678, 910111213), uuid_2);
+    EXPECT_NE(variable1.uuid(), variable2.uuid());
+  }
+
+  // Verify two orientations with the same hardware ID and different timestamps produce different UUIDs
+  {
+    Orientation3DStamped variable1(ros::Time(12345678, 910111213), uuid_1);
+    Orientation3DStamped variable2(ros::Time(12345678, 910111214), uuid_1);
+    EXPECT_NE(variable1.uuid(), variable2.uuid());
+
+    Orientation3DStamped variable3(ros::Time(12345678, 910111213), uuid_1);
+    Orientation3DStamped variable4(ros::Time(12345679, 910111213), uuid_1);
+    EXPECT_NE(variable3.uuid(), variable4.uuid());
+  }
+
+  // Verify two orientations with different hardware IDs and different timestamps produce different UUIDs
+  {
+    Orientation3DStamped variable1(ros::Time(12345678, 910111213), uuid_1);
+    Orientation3DStamped variable2(ros::Time(12345678, 910111214), uuid_2);
+    EXPECT_NE(variable1.uuid(), variable2.uuid());
+
+    Orientation3DStamped variable3(ros::Time(12345678, 910111213), uuid_1);
+    Orientation3DStamped variable4(ros::Time(12345679, 910111213), uuid_2);
+    EXPECT_NE(variable3.uuid(), variable4.uuid());
+  }
+}
+
+struct QuaternionCostFunction
+{
+  explicit QuaternionCostFunction(double *observation)
+  {
+    observation_[Orientation3DStamped::W] = observation[Orientation3DStamped::W];
+    observation_[Orientation3DStamped::X] = observation[Orientation3DStamped::X];
+    observation_[Orientation3DStamped::Y] = observation[Orientation3DStamped::Y];
+    observation_[Orientation3DStamped::Z] = observation[Orientation3DStamped::Z];
+  }
+
+  bool operator()(const double* quaternion, double* residual) const
+  {
+    tf2::Quaternion quat1(
+      quaternion[Orientation3DStamped::X],
+      quaternion[Orientation3DStamped::Y],
+      quaternion[Orientation3DStamped::Z],
+      quaternion[Orientation3DStamped::W]);
+
+    tf2::Quaternion quat2(
+      observation_[Orientation3DStamped::X],
+      observation_[Orientation3DStamped::Y],
+      observation_[Orientation3DStamped::Z],
+      observation_[Orientation3DStamped::W]);
+
+    tf2::Quaternion diff = quat2;
+    diff *= quat1.inverse();
+
+    residual[0] = diff.x();
+    residual[1] = diff.y();
+    residual[2] = diff.z();
+
+    return true;
+  }
+
+  double observation_[4];
+};
+
+struct QuaternionCostFunctor
+{
+  explicit QuaternionCostFunctor(double* observation) :
+    function_to_functor_(
+      new ceres::NumericDiffCostFunction<QuaternionCostFunction, ceres::CENTRAL, 3, 4>(
```

Comment:
I'm dying to know if the auto-diff version gets closer to a normalized quaternion than the numerical-diff version.

Reply:
I'm guessing switching to it will require more than changing this to the other constructor. I'll give it a go.